File: smoke_tests/dart/path_traversal.dart
// Path Traversal vulnerabilities in Dart

import 'dart:io';

// Test 1: File read with user path
Future<String> vulnerableFileRead(String userPath) async {
  // VULNERABLE: Path traversal via user input
  var file = File(userPath);
  return await file.readAsString();
}

// Test 2: File read with base path concatenation
Future<String> vulnerablePathConcat(String filename) async {
  // VULNERABLE: Path traversal via concatenation
  var file = File('/var/data/' + filename);
  return await file.readAsString();
}

// Test 3: File write with user path
Future<void> vulnerableFileWrite(String userPath, String content) async {
  // VULNERABLE: Arbitrary file write
  var file = File(userPath);
  await file.writeAsString(content);
}

// Test 4: File delete with user path
Future<void> vulnerableFileDelete(String userPath) async {
  // VULNERABLE: Arbitrary file deletion
  var file = File(userPath);
  await file.delete();
}

// Test 5: Directory listing with user path
Future<List<FileSystemEntity>> vulnerableDirList(String userDir) async {
  // VULNERABLE: Directory traversal
  var dir = Directory(userDir);
  return await dir.list().toList();
}

// Test 6: Directory creation with user path
Future<void> vulnerableDirCreate(String userDir) async {
  // VULNERABLE: Arbitrary directory creation
  var dir = Directory(userDir);
  await dir.create(recursive: true);
}

// Test 7: File copy with user paths
Future<void> vulnerableFileCopy(String src, String dst) async {
  // VULNERABLE: Path traversal in copy
  var srcFile = File(src);
  await srcFile.copy(dst);
}

// Test 8: File rename/move with user paths
Future<void> vulnerableFileRename(String oldPath, String newPath) async {
  // VULNERABLE: Path traversal in rename
  var file = File(oldPath);
  await file.rename(newPath);
}

// Test 9: Symlink creation with user paths
Future<void> vulnerableSymlink(String target, String link) async {
  // VULNERABLE: Symlink attack
  var linkEntity = Link(link);
  await linkEntity.create(target);
}

// Test 10: File path from query parameter
Future<String> vulnerableQueryParam(Uri uri) async {
  // VULNERABLE: Path from query parameter
  var filename = uri.queryParameters['file'];
  if (filename != null) {
    return await File('/uploads/$filename').readAsString();
  }
  return '';
}

// Test 11: Archive extraction with user path
Future<void> vulnerableArchiveExtract(String archivePath, String destDir) async {
  // VULNERABLE: Zip slip
  await Process.run('unzip', [archivePath, '-d', destDir]);
}

// Test 12: File from form data
Future<void> vulnerableFormUpload(String uploadedFilename, List<int> data) async {
  // VULNERABLE: User-controlled filename
  var file = File('/uploads/$uploadedFilename');
  await file.writeAsBytes(data);
}

// Test 13: Path with URL decoding
Future<String> vulnerableUrlDecoded(String encodedPath) async {
  // VULNERABLE: URL-decoded path
  var decodedPath = Uri.decodeComponent(encodedPath);
  return await File(decodedPath).readAsString();
}

// Test 14: FileSystemEntity.type with user path
Future<void> vulnerableFileType(String path) async {
  // VULNERABLE: Information disclosure
  var type = await FileSystemEntity.type(path);
  print('Type: $type');
}

// Test 15: RandomAccessFile with user path
Future<void> vulnerableRandomAccess(String path) async {
  // VULNERABLE: Direct file access
  var file = await File(path).open(mode: FileMode.read);
  var contents = await file.read(1024);
  await file.close();
}

File: smoke_tests/perl/path_traversal.pl
#!/usr/bin/perl
# Path Traversal vulnerabilities in Perl

use strict;
use warnings;
use CGI;
use File::Spec;

my $cgi = CGI->new();

# Test 1: open() with user input
sub vulnerable_open {
    my $filename = $cgi->param('file');
    # VULNERABLE: Path traversal via open
    open(my $fh, '<', $filename) or die "Cannot open: $!";
    my @content = <$fh>;
    close($fh);
    return @content;
}

# Test 2: sysopen with user input
sub vulnerable_sysopen {
    my $file = $cgi->param('file');
    # VULNERABLE: Path traversal via sysopen
    sysopen(my $fh, $file, O_RDONLY) or die "Cannot open: $!";
    my $content;
    sysread($fh, $content, 1024);
    close($fh);
    return $content;
}

# Test 3: open for writing
sub vulnerable_open_write {
    my $filename = $cgi->param('filename');
    my $content = $cgi->param('content');
    # VULNERABLE: Arbitrary file write
    open(my $fh, '>', "/var/www/uploads/$filename") or die "Cannot write: $!";
    print $fh $content;
    close($fh);
}

# Test 4: unlink with user input
sub vulnerable_unlink {
    my $file = $cgi->param('file');
    # VULNERABLE: Arbitrary file deletion
    unlink("/var/www/uploads/$file");
}

# Test 5: rename with user input
sub vulnerable_rename {
    my $old = $cgi->param('old');
    my $new = $cgi->param('new');
    # VULNERABLE: Path traversal in rename
    rename("/var/www/files/$old", "/var/www/files/$new");
}

# Test 6: symlink creation
sub vulnerable_symlink {
    my $target = $cgi->param('target');
    my $link = $cgi->param('link');
    # VULNERABLE: Symlink attack
    symlink($target, "/var/www/files/$link");
}

# Test 7: mkdir with user input
sub vulnerable_mkdir {
    my $dir = $cgi->param('dir');
    # VULNERABLE: Directory traversal
    mkdir("/var/www/data/$dir", 0755);
}

# Test 8: rmdir with user input
sub vulnerable_rmdir {
    my $dir = $cgi->param('dir');
    # VULNERABLE: Directory traversal in delete
    rmdir("/var/www/data/$dir");
}

# Test 9: chmod with user input
sub vulnerable_chmod {
    my $file = $cgi->param('file');
    my $mode = $cgi->param('mode');
    # VULNERABLE: Arbitrary file permission change
    chmod(oct($mode), "/var/www/files/$file");
}

# Test 10: chown with user input
sub vulnerable_chown {
    my $file = $cgi->param('file');
    my $uid = $cgi->param('uid');
    my $gid = $cgi->param('gid');
    # VULNERABLE: Arbitrary file ownership change
    chown($uid, $gid, "/var/www/files/$file");
}

# Test 11: Link creation
sub vulnerable_link {
    my $source = $cgi->param('source');
    my $dest = $cgi->param('dest');
    # VULNERABLE: Hard link creation
    link("/var/www/files/$source", "/var/www/files/$dest");
}

# Test 12: truncate with user input
sub vulnerable_truncate {
    my $file = $cgi->param('file');
    # VULNERABLE: File truncation
    truncate("/var/www/files/$file", 0);
}

# Test 13: File::Slurp read
sub vulnerable_file_slurp {
    use File::Slurp;
    my $file = $cgi->param('file');
    # VULNERABLE: Path traversal
    my $content = read_file($file);
    return $content;
}

# Test 14: readdir abuse
sub vulnerable_readdir {
    my $dir = $cgi->param('dir');
    # VULNERABLE: Directory listing outside intended path
    opendir(my $dh, $dir) or die "Cannot open directory: $!";
    my @files = readdir($dh);
    closedir($dh);
    return @files;
}

# Test 15: File::Copy operations
sub vulnerable_file_copy {
    use File::Copy;
    my $source = $cgi->param('source');
    my $dest = $cgi->param('dest');
    # VULNERABLE: Path traversal in copy
    copy("/var/www/files/$source", "/var/www/backup/$dest");
}

1;
